crates/ide-diagnostics/src/handlers/remove_trailing_return.rs

```diff
@@ -0,0 +1,377 @@
+use hir::{db::ExpandDatabase, diagnostics::RemoveTrailingReturn, HirFileIdExt, InFile};
+use ide_db::{assists::Assist, source_change::SourceChange};
+use syntax::{ast, AstNode, SyntaxNodePtr};
+use text_edit::TextEdit;
+
+use crate::{fix, Diagnostic, DiagnosticCode, DiagnosticsContext};
+
+// Diagnostic: remove-trailing-return
+//
+// This diagnostic is triggered when there is a redundant `return` at the end of a function
+// or closure.
+pub(crate) fn remove_trailing_return(
+    ctx: &DiagnosticsContext<'_>,
+    d: &RemoveTrailingReturn,
+) -> Diagnostic {
+    let display_range = ctx.sema.diagnostics_display_range(InFile {
+        file_id: d.file_id,
+        value: expr_stmt(ctx, d)
+            .as_ref()
+            .map(|stmt| SyntaxNodePtr::new(stmt.syntax()))
+            .unwrap_or_else(|| d.return_expr.into()),
+    });
+    Diagnostic::new(
+        DiagnosticCode::Clippy("needless_return"),
+        "replace return <expr>; with <expr>",
+        display_range,
+    )
+    .with_fixes(fixes(ctx, d))
+}
+
+fn fixes(ctx: &DiagnosticsContext<'_>, d: &RemoveTrailingReturn) -> Option<Vec<Assist>> {
+    let return_expr = return_expr(ctx, d)?;
+    let stmt = expr_stmt(ctx, d);
+
+    let range = stmt.as_ref().map_or(return_expr.syntax(), AstNode::syntax).text_range();
```

Comment:
This range might be pointing into a macro expansion, we need to map that up
```diff
    let FileRange { range, file_id } = ctx.sema.original_range_opt(stmt.as_ref().map_or(return_expr.syntax(), AstNode::syntax))?;
```
and then use or check that this file_id is the same as the diagnostic one (only case where this could differ is from includes)